### ISSUE TITLE
FontForge generates files with both GPOS and `kern` tables, which don't have correct kerning and need post-processing

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3435]
This is related, I believe, to #3415.

FontForge generates fonts which have no working kerning unless I manually remove the "GPOS" table that FontForge generates for some reason and which seemingly can't be disabled.

For example, after I follow all but the last step [my build process](https://github.com/ctrlcctrlv/kjv1611/blob/master/CONTRIBUTING.md)...

If I were to open the generated OTF (attached) with FontForge, it will spit out a warning:
```
This font contains both a 'kern' table and a 'GPOS' table.
  The 'kern' table will only be read if there is no 'kern' feature in 'GPOS'.
```

The generated file's kerning will also be completely broken (note the word "fox").

If I do the last step in [my build process](https://github.com/ctrlcctrlv/kjv1611/blob/master/CONTRIBUTING.md) which eliminates the GPOS table, it will work fine.

![fox469](https://user-images.githubusercontent.com/838783/51984252-d7e12000-24d5-11e9-9a4b-c7f6932ff0fe.png)

* Why does the GPOS table cause this?
* How can I direct FontForge to not generate GPOS but still generate kern?
* Can FontForge be fixed not to cause this?

@skef requested I make separate issue.

**Comment by skef** [comment]
When I follow your instructions but skip the last step, I get the `.otf` in this archive that looks quite different from the one stored in your repo: [KJV.zip](https://github.com/fontforge/fontforge/files/2814772/KJV.zip). Note that I'm picking "OpenType (CCF)" from the Generate drop-down. I get an (effectively) identical file if I uncomment the last two lines in `script.txt`. According to `ttx` these don't have a separate `kern` table and don't generate the warning when opened in FontForge. 

Since I'm using a version of FontForge built from a recent copy of the master, and I believe you've also been compiling recent versions, it seems like the next step is trying to determine why the files are so different. What is the last commit on your version, the OS of your system, and how are you building?

**Comment by ctrlcctrlv** [comment]
@skef Hi, I should have included my non-default export parameters. Try again using them :-)
![buildopts](https://user-images.githubusercontent.com/838783/52032906-fc80da80-255d-11e9-8a84-ce04a4cafe62.png)

**Comment by ctrlcctrlv** [comment]
>Any tool that can use GSUB but not GPOS is at least a little suspicious.

FreeType is such a tool, per [FreeType 2 Tutorial
Step 2 — managing glyphs](https://tools.ietf.org/doc/libfreetype6/tutorial/step2.html), section 3 subsection c.

**Comment by skef** [comment]
The issue is not just the potential for needing `kern` and `GSUB`, as FontForge already allows that combination. It is a tool that knows about `GSUB`, knows about and prefers `kern`, but fails to work if `GPOS` is also included. 

Are you sure that FreeType 2 handles GSUB? The website implies 1 did but I didn't find any positive statement that 2 does, and searching turns up a few anecdotal claims that it doesn't, or doesn't directly.

**Comment by ctrlcctrlv** [comment]
I'm sorry, I was confused by the weird way that FontForge chooses to show this to users. I thought that `liga` was part of `GSUB` because that's how FontForge displays it, but that's not the case.

**Comment by skef** [comment]
`liga` is part of GSUB, yes, and I don't believe there is a standard older alternative *substitution* map. But Unicode has code points for a limited number of ligatures and your font uses at least some of them. `f_i` is mapped to `U+fb01` for example. That could explain why you're seeing some of your ligatures work without GSUB (if that's the datum needing explanation).

**Comment by ctrlcctrlv** [comment]
If I disable OpenType exports I lose my  Historical Forms (`hist`) table along with many other necessary OpenType tables so unfortunately I cannot do that. And no, I'm not talking about just `f_i`, I have a ligature in that font `L O R D` which works in GIMP, [which uses FreeType](https://docs.gimp.org/2.10/nl/gimp-using-fonts.html). Even my Contextual Alternates (`calt`) feature is working.

![gsub](https://user-images.githubusercontent.com/838783/52039791-715d1000-2570-11e9-86cf-8e855091a567.png)

In plaintext that's:

```
LORD
L O R D
fwswſwwfox
```

**Comment by ctrlcctrlv** [comment]
I think if `kern` is checked, `GPOS` should still be generated, but shouldn't itself include `kern`&mdash;that should fix this problem if I read FontForge's warning correctly. My hack, which removes the whole table, is too heavy-handed for a general solution.

**Comment by skef** [comment]
Gimp uses FreeType with Pango and Pango has OpenType feature support. 

> I think if `kern` is checked, `GPOS` should still be generated, but shouldn't itself include `kern`—that should fix this problem if I read FontForge's warning correctly. 

That isn't a general solution. There are going to be many OpenType implementations that won't use `kern`. This problem with FreeType *could* warrant a separate flag to handle the case, but just deleting `kern` out of `GPOS` because FreeType 2 can't handle it being there is very likely to break other cases.

**Comment by ctrlcctrlv** [comment]
OK, sorry.

By the way, just noticed from my example above that some of my kerns aren't working even after post-processing...the sequence `f w`, which becomes `f w.alt` under `calt` should be kerned. Should I make a new issue for this?

**Comment by ctrlcctrlv** [comment]
@skef Awesome, I gave all my glyphs PUA encodings and it works now. You rock.

![kerning alternates](https://user-images.githubusercontent.com/838783/52045129-c4d65a80-257e-11e9-8305-7c245089885e.png)

The main issue still stands though.

**Comment by skef** [comment]
My sense is that the number of users who

1. Would benefit from such a feature
2. Could be informed that they would benefit from such a feature, and
3. Wouldn't be just as well served by a work-around

is probably vanishingly small. 

You have a partial work-around already except that you are unhappy that all of `GPOS` goes away. But you don't currently have any other GPOS types in your font. So is this a generic worry about other potential users or are you worried about needing to add other types in the future?

**Comment by skef** [comment]
(BTW: the reason only mapped glyphs are winding up in your `kern` table is because that's what the `windows` checkbox in the dialog does.)

**Comment by ctrlcctrlv** [comment]
Hi&mdash;I reported this bug as a generic worry. I believe it's counterintuitive to require users use hacks just to get a font that behaves well with both FreeType and MS Windows when FontForge could generate a font itself with the appropriate options. We have tick boxes for "Apple" and a bunch of other platform-specific generation options so I don't see why this can't be added...

**Comment by skef** [comment]
Did you/can you confirm that an OTF generated without a `kern` table doesn't kern correctly in GIMP?

**Comment by ctrlcctrlv** [comment]
Yes. No `kern` table, no kerning. In this example, the first font was generated with `kern`, the second only with "OpenType". `kern` is required to have kerning, GPOS alone is not enough.

![kern](https://user-images.githubusercontent.com/838783/52050644-2e109a80-258c-11e9-9d44-449f66610099.png)

**Comment by skef** [comment]
I'm just a random person doing stuff here for a bit, so it's not really my business what and how issues should be addressed. Still, I've been looking at this issue so I'll summarize what further information I would need in order to go ahead with a fix.

Any added flag should have an associated explanation for when it might be used. @ctrlcctrlv 's conception of its purpose hasn't yet accounted for all of the evidence. It can't really be a flag for FreeType 2 compatibility because FreeType 2 doesn't include either GSUB or GPOS processing. Roughly speaking, the way to be FreeType 2 compatible is to be more or less a TrueType font. 

Therefore what this flag would address is software that:

1. Has at least some `GSUB` support,
2. Does not use `GPOS` (for kerning, at least), and
3. For whatever reason will not use an embedded `kern` table if `GPOS` kerning is present. 

(Note: if the proposal is to output non-`kern` `GPOS` tables, before going forward we would need to confirm that it's just `kern` and not any `GPOS` tables that cause the problem.) 

GIMP is confirmed as one such application, and is significant enough an application that it might warrant a workaround. 

However, suppose that GIMP were the only such application we knew of, and the proposal were to add a    "GIMP compatibility" checkbox. I would not work on such a fix before pursuing the problem with the GIMP development team. Indeed, I would much rather spend time digging around a bit in their code than making an exception in FontForge. New and unusual font format combinations shouldn't be added into the ecosystem lightly -- they tend to stick around. I cringe at the thought of later projects needing to add support for a `GSUB`/`kern` combination because FontForge made it easy to build such fonts and they wound up in distributions that are unable or uninterested in rebuilding them. 

This problem could easily be wider than GIMP, of course. But in that case we still have a documentation issue -- how do we indicate to a user what the flag is for. And if the answer points to a particular library or combination of libraries the same question arises: Why not file a bug against the relevant library instead? 
We *know* it's not a hard fix because all that needs to happen is that `GSUB` kerning entries be ignored. 

In short, the missing information is the scope of the problem, and once determined the scope of the problem may point away from changes to FontForge itself.

**Comment by skef** [comment]
Given that Pango appears to depend on HarfBuzz and likely uses it for GSUB access, the code referenced in HarfBuzz/HarfBuzz#250 is likely to be near the problem if it is indeed attributable to Pango.

**Comment by skef** [comment]
The, I guess, "intermediate" point here is that this issue is about working around the quirks of one or more particular implementations. If those implementations can deal with unified GPOS kern tables then that is a better workaround than simply deleting them. 

It's easy to imagine implementation tricks that would support multiple tables *in general* while not supporting particular multi-table implementations. It could be that the problem is multiple tables with all the same script/language entries. Or maybe the current implementation is just a "best effort" attempt.

**Comment by khaledhosny** [comment]
Multiple subtables are supported, but multiple subtables where character coverage overlap will not do what you expect; since a character that is covered in one subtable will be skipped for all subsequent subtables. Either merge the subtables or split them into different lookups.

There is no FontForge issue here, FontForge is compiling the GPOS lookups as you created them faithfully, but your lookups don’t do what you think they are doing. I doubt that you will find any implementation where your GPOS kerning works, but even if you do it is still not a FontForge issue.

For better or worth, FontForge representation of OpenType lookup is very close to the metal with little abstraction, this can make it clumsy to use at times, but on the other hand you can create pretty much any lookup supported by the standard (which is not the case for some other higher level representations).

**Comment by ctrlcctrlv** [comment]
I'm confused. On Linux, Freetype is used and it doesn't support GPOS. So why should it matter that my generated GPOS tables are incorrect? Aren't all applications ignoring them anyway since the vast majority use Freetype to render fonts?

**Comment by khaledhosny** [comment]
You are indeed confused :slightly_smiling_face:

FreeType does not do text layout, it does rasterization (some simple/ancient applications will use it for text layout, but these are usually broken in other ways as they won’t supports GSUB or complex text layout, so pretty irrelevant). 

On Linux libraries/applications that do text layout (e.g. Pango, Qt, Firefox, Chrome etc.) will use HarfBuzz either directly or indirectly. HarfBuzz supports GPOS kerning of course and will use old-style kern table only if the fonts does not have a a GPOS table.

**Comment by ctrlcctrlv** [comment]
Just hit this issue again https://github.com/ctrlcctrlv/kjv1611/issues/33#issuecomment-465937222

**Comment by khaledhosny** [comment]
Again, your kerning lookups are wrongly setup. Either merge your subtables into one, or move each subtable into a separate lookup.  Nothing to be fixed in FontForge, it is correctly compiling your lookups.

**Comment by ctrlcctrlv** [comment]
>move each subtable into a separate lookup

Let me try this :+1:

**Comment by ctrlcctrlv** [comment]
It worked! :tada: Thank you so much @khaledhosny
